network/p2p/tracer/internal/rpc_sent_tracker.go

```diff
@@ -1,32 +1,130 @@
 package internal
 
 import (
+	"crypto/rand"
+	"fmt"
+	"time"
+
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
 	pb "github.com/libp2p/go-libp2p-pubsub/pb"
 	"github.com/rs/zerolog"
+	"go.uber.org/atomic"
 
+	"github.com/onflow/flow-go/engine/common/worker"
 	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/mempool/queue"
 	p2pmsg "github.com/onflow/flow-go/network/p2p/message"
 )
 
-// RPCSentTracker tracks RPC messages that are sent.
+// trackableRPC is an internal data structure for "temporarily" storing *pubsub.RPC sent in the queue before they are processed
+// by the *RPCSentTracker.
+type trackableRPC struct {
+	// Nonce prevents deduplication in the hero store
+	Nonce []byte
+	rpc   *pubsub.RPC
+}
+
+// lastHighestIHaveRPCSize tracks the last highest rpc control message size the time stamp it was last updated.
+type lastHighestIHaveRPCSize struct {
+	*atomic.Int64
+	lastUpdate time.Time
+}
```

Comment:
This entire structure must be concurrency-safe, currently only the size part is atomic, and the time is non-concurrency safe.

```go
// lastHighestIHaveRPCSize tracks the last highest rpc control message size the time stamp it was last updated.
type lastHighestIHaveRPCSize struct {
	sync.Mutex
	lastSize   int
	lastUpdate time.Time
}

// update updates the last highest ihave RPC size atomically.
// If the new size is larger than the current size, or if the time elapsed since the last update surpasses the reset interval.
// The time-based reset interval is used to prevent the last highest ihave RPC size from being permanently stuck at a very large size.
func (l *lastHighestIHaveRPCSize) update(size int, ttl time.Duration) {
	l.Lock()
	defer l.Unlock()

	if l.lastSize < size || time.Since(l.lastUpdate) > ttl {
		// The last highest ihave RPC size is updated if the new size is larger than the current size, or if the time elapsed since the last update surpasses the reset interval.
		l.lastSize = size
	}
}
```

Reply:
should this also update `lastUpdate` within the lock? doesn't look like it's set anywhere

Reply:
You are right, updated it.

Reply:
https://github.com/onflow/flow-go/pull/4553/commits/999eb9bdeaf15b420e76b8d76546dfd62f445611